File: mockserver-client-php/tests/Unit/RoundTripFidelityTest.php
<?php

declare(strict_types=1);

namespace MockServer\Tests\Unit;

use MockServer\Expectation;
use PHPUnit\Framework\Attributes\DataProvider;
use PHPUnit\Framework\TestCase;

/**
 * Cross-language JSON round-trip fidelity harness for the PHP client.
 *
 * For every shared fixture in {@code test-fixtures/expectations/*.json} this
 * deserializes with the client model ({@see Expectation::fromArray()}),
 * re-serializes ({@code json_encode()} via {@see Expectation::jsonSerialize()})
 * and asserts the round-trip is semantically equal to the input, modulo a
 * known-gaps ledger.
 *
 * The comparator (NORM + CANON + DIFFS + EXCUSED) is a faithful port of the
 * shared reference in {@code .tmp/reference_compare.py} so every language port
 * produces the same diff paths for the same client behaviour.
 *
 * The PHP client stores the decoded array verbatim in {@code rawData} and
 * replays it unchanged from {@code toArray()}, so PHP is expected to round-trip
 * every field with ZERO gaps.
 */
final class RoundTripFidelityTest extends TestCase
{
    private const LANG = 'php';

    /** keyToMultiValue fields (headers/params/trailers dual encoding). */
    private const MULTI = ['headers', 'queryStringParameters', 'trailers'];

    /** keyToValue fields (cookies dual encoding). */
    private const SINGLE = ['cookies'];

    private static function repoRoot(): string
    {
        // tests/Unit -> tests -> mockserver-client-php -> repo root
        return __DIR__ . '/../../../';
    }

    /**
     * @return list<string> absolute fixture paths, excluding the manifest.
     */
    private static function fixtureFiles(): array
    {
        $glob = self::repoRoot() . 'test-fixtures/expectations/*.json';
        $files = glob($glob);
        if ($files === false) {
            return [];
        }
        $files = array_filter(
            $files,
            static fn(string $f): bool => basename($f) !== 'known-gaps.json',
        );
        sort($files);
        return array_values($files);
    }

    /**
     * @return array<string, mixed> the full known-gaps manifest.
     */
    private static function manifest(): array
    {
        $override = getenv('FIDELITY_KNOWN_GAPS');
        $path = ($override !== false && $override !== '')
            ? $override
            : self::repoRoot() . 'test-fixtures/expectations/known-gaps.json';
        $json = file_get_contents($path);
        if ($json === false) {
            return [];
        }
        $decoded = json_decode($json, true);
        return is_array($decoded) ? $decoded : [];
    }

    /**
     * @return list<string> excused path entries for this language.
     */
    private static function gaps(): array
    {
        $manifest = self::manifest();
        $entries = $manifest[self::LANG] ?? [];
        return is_array($entries) ? array_values(array_filter(
            $entries,
            static fn($e): bool => is_string($e),
        )) : [];
    }

    // ------------------------------------------------------------------
    // Comparator port (NORM + CANON + DIFFS + EXCUSED) over assoc arrays.
    // ------------------------------------------------------------------

    private static function isJsonObject(mixed $v): bool
    {
        return is_array($v) && !array_is_list($v);
    }

    private static function isJsonList(mixed $v): bool
    {
        return is_array($v) && array_is_list($v);
    }

    /**
     * @return array<string, list<mixed>> canonical {name -> [values...]} map.
     */
    private static function canonMulti(mixed $v): array
    {
        $out = [];
        if (self::isJsonObject($v)) {
            foreach ($v as $k => $val) {
                $out[$k] = self::isJsonList($val) ? $val : [$val];
            }
        } elseif (self::isJsonList($v)) {
            foreach ($v as $e) {
                if (is_array($e) && array_key_exists('name', $e)) {
                    if (array_key_exists('values', $e)) {
                        $vals = $e['values'];
                    } else {
                        $vals = array_key_exists('value', $e) ? $e['value'] : null;
                    }
                    $out[$e['name']] = self::isJsonList($vals) ? $vals : [$vals];
                }
            }
        }
        return $out;
    }

    /**
     * @return array<string, mixed> canonical {name -> value} map.
     */
    private static function canonSingle(mixed $v): array
    {
        $out = [];
        if (self::isJsonObject($v)) {
            $out = $v;
        } elseif (self::isJsonList($v)) {
            foreach ($v as $e) {
                if (is_array($e) && array_key_exists('name', $e)) {
                    $out[$e['name']] = array_key_exists('value', $e) ? $e['value'] : null;
                }
            }
        }
        return $out;
    }

    /**
     * NORM (null==absent) + CANON (dual-encoding) in one pass, keyed by parent
     * key name.
     */
    private static function norm(mixed $v, ?string $key = null): mixed
    {
        if ($v === null) {
            return null;
        }
        if ($key !== null && in_array($key, self::MULTI, true)) {
            $r = [];
            foreach (self::canonMulti($v) as $k => $vs) {
                $r[$k] = array_map(static fn($x) => self::norm($x), $vs);
            }
            return $r;
        }
        if ($key !== null && in_array($key, self::SINGLE, true)) {
            $r = [];
            foreach (self::canonSingle($v) as $k => $x) {
                $r[$k] = self::norm($x);
            }
            return $r;
        }
        if (self::isJsonList($v)) {
            return array_map(static fn($x) => self::norm($x), $v);
        }
        if (is_array($v)) { // object
            $r = [];
            foreach ($v as $k => $x) {
                if ($x !== null) {
                    $r[(string) $k] = self::norm($x, (string) $k);
                }
            }
            return $r;
        }
        return $v;
    }

    /**
     * @return list<string> diff path strings (input `a` vs output `b`).
     */
    private static function diffs(mixed $a, mixed $b, string $path = ''): array
    {
        $res = [];
        if (self::isJsonObject($a)) {
            if (!self::isJsonObject($b)) {
                return [$path !== '' ? $path : '<root>'];
            }
            foreach ($a as $k => $v) {
                $p = $path !== '' ? "$path.$k" : (string) $k;
                if (!array_key_exists($k, $b)) {
                    $res[] = $p;
                } else {
                    $res = array_merge($res, self::diffs($v, $b[$k], $p));
                }
            }
            foreach ($b as $k => $_) {
                if (!array_key_exists($k, $a)) {
                    $res[] = ($path !== '' ? "$path.$k" : (string) $k) . ' [ADDED]';
                }
            }
            return $res;
        }
        if (self::isJsonList($a)) {
            if (!self::isJsonList($b)) {
                return [$path !== '' ? $path : '<root>'];
            }
            $bCount = count($b);
            foreach ($a as $i => $v) {
                $p = $path !== '' ? "$path.$i" : (string) $i;
                if ($i >= $bCount) {
                    $res[] = $p;
                } else {
                    $res = array_merge($res, self::diffs($v, $b[$i], $p));
                }
            }
            return $res;
        }
        // scalar (or empty-array vs empty-array, both isList -> handled above)
        if (!self::scalarEqual($a, $b)) {
            $res[] = $path !== '' ? $path : '<root>';
        }
        return $res;
    }

    /**
     * Scalar equality matching the reference comparator's Python `!=` semantics.
     *
     * Python compares numbers by value across int/float ({@code 5.0 == 5}), but
     * treats cross-kind scalars (string vs number, etc.) as unequal. PHP's
     * {@code json_encode} renders a whole-number float such as {@code 5.0} as
     * {@code 5}, so after the JSON round-trip a fixture's {@code 5.0} decodes
     * back as the int {@code 5}. That is a representation change, not a semantic
     * one, and the shared comparator (per the Python reference) considers the
     * two numerically-equal scalars equal — so it must NOT be flagged as a diff.
     * Non-numeric scalars use strict comparison to avoid PHP's loose-`==`
     * coercions (e.g. {@code "1" == 1}) that Python does not perform.
     */
    private static function scalarEqual(mixed $a, mixed $b): bool
    {
        $aNum = is_int($a) || is_float($a);
        $bNum = is_int($b) || is_float($b);
        if ($aNum && $bNum) {
            return $a == $b;
        }
        return $a === $b;
    }

    private static function star(string $p): string
    {
        return implode('.', array_map(
            static fn(string $s): string => ctype_digit($s) ? '*' : $s,
            explode('.', $p),
        ));
    }

    /**
     * @param list<string> $entries
     */
    private static function excused(string $path, array $entries): bool
    {
        $pSegs = explode('.', $path);
        foreach ($entries as $g) {
            $gSegs = explode('.', $g);
            if (count($gSegs) > count($pSegs)) {
                continue;
            }
            $ok = true;
            foreach ($gSegs as $i => $gSeg) {
                $pSeg = $pSegs[$i];
                if ($gSeg === $pSeg) {
                    continue;
                }
                if ($gSeg === '*' && ctype_digit($pSeg)) {
                    continue;
                }
                $ok = false;
                break;
            }
            if ($ok) {
                return true;
            }
        }
        return false;
    }

    /**
     * Round-trip a fixture and return the raw diff paths (with any `[ADDED]`
     * suffix preserved).
     *
     * @return list<string>
     */
    private static function computeDiffs(string $fixturePath): array
    {
        $json = file_get_contents($fixturePath);
        if ($json === false) {
            throw new \RuntimeException("Cannot read fixture: $fixturePath");
        }
        $input = json_decode($json, true, 512, JSON_THROW_ON_ERROR);

        // Deserialize with the client model, then re-serialize via json_encode.
        $expectation = Expectation::fromArray($input);
        $reencoded = json_encode($expectation, JSON_THROW_ON_ERROR);
        $output = json_decode($reencoded, true, 512, JSON_THROW_ON_ERROR);

        return self::diffs(self::norm($input), self::norm($output));
    }

    // ------------------------------------------------------------------
    // Data provider — one data set per fixture, named by filename.
    // ------------------------------------------------------------------

    /**
     * @return iterable<string, array{string}>
     */
    public static function fixtureProvider(): iterable
    {
        foreach (self::fixtureFiles() as $path) {
            yield basename($path) => [$path];
        }
    }

    /**
     * Every fixture must round-trip with no unexcused diffs.
     */
    #[DataProvider('fixtureProvider')]
    public function testRoundTripFidelity(string $fixturePath): void
    {
        $gaps = self::gaps();
        $unexcused = [];
        foreach (self::computeDiffs($fixturePath) as $d) {
            $bare = str_replace(' [ADDED]', '', $d);
            if (!self::excused($bare, $gaps)) {
                $unexcused[] = $d;
            }
        }

        $this->assertSame(
            [],
            $unexcused,
            'Unexcused round-trip fidelity diffs for ' . basename($fixturePath)
            . ': ' . implode(', ', $unexcused),
        );
    }

    /**
     * Confirms the fixture set is the expected size (guards against a broken
     * glob / path silently testing nothing).
     */
    public function testFixtureCount(): void
    {
        $this->assertCount(44, self::fixtureFiles(), 'Expected 44 fixtures (excluding known-gaps.json)');
    }

    /**
     * Ratchet: every entry in gaps[php] must excuse at least one real diff path
     * across all fixtures. A stale entry means the model was fixed and the
     * manifest entry should be removed — CI fails until it is.
     *
     * (Trivially passes for an empty gap list, which is the expected PHP state.)
     */
    public function testNoStaleGapEntries(): void
    {
        $gaps = self::gaps();
        if ($gaps === []) {
            $this->assertSame([], $gaps);
            return;
        }

        $allDiffs = [];
        foreach (self::fixtureFiles() as $path) {
            foreach (self::computeDiffs($path) as $d) {
                $allDiffs[] = str_replace(' [ADDED]', '', $d);
            }
        }

        $stale = [];
        foreach ($gaps as $entry) {
            $used = false;
            foreach ($allDiffs as $diff) {
                if (self::excused($diff, [$entry])) {
                    $used = true;
                    break;
                }
            }
            if (!$used) {
                $stale[] = $entry;
            }
        }

        $this->assertSame(
            [],
            $stale,
            'Stale gaps[php] entries excusing no diff (remove them): ' . implode(', ', $stale),
        );
    }

    /**
     * Discovery mode: when FIDELITY_DISCOVER=1, print the sorted, deduped,
     * `*`-index-normalized union of computed diff paths for PHP and pass. This
     * is how the orchestrator harvests the language gap list.
     */
    public function testDiscoverGaps(): void
    {
        if (getenv('FIDELITY_DISCOVER') !== '1') {
            $this->markTestSkipped('Set FIDELITY_DISCOVER=1 to print the computed PHP gap list.');
        }

        $union = [];
        foreach (self::fixtureFiles() as $path) {
            foreach (self::computeDiffs($path) as $d) {
                $union[self::star(str_replace(' [ADDED]', '', $d))] = true;
            }
        }
        $paths = array_keys($union);
        sort($paths);

        fwrite(STDOUT, "\nFIDELITY_DISCOVER php gaps:\n" . json_encode($paths, JSON_PRETTY_PRINT) . "\n");
        $this->assertTrue(true);
    }
}

File: mockserver-client-go/roundtrip_fidelity_test.go
package mockserver

// Cross-language client fidelity harness — GO port.
//
// For every fixture in repo-root test-fixtures/expectations/*.json (excluding
// the known-gaps.json manifest) this test unmarshals the JSON into the
// Expectation struct and re-marshals it, then compares the input and output as
// generic JSON values under a shared comparator (NORM + CANON + DIFFS) that is
// implemented identically across all client languages. Any path that differs is
// a fidelity gap (Go's pure encoding/json model silently drops unknown keys).
// Gaps are excused via the "go" entry in the known-gaps.json ledger; the ratchet
// subtest fails if a ledger entry no longer excuses any diff (i.e. the model was
// fixed and the entry is now stale).
//
// The exact comparator algorithm is specified in .tmp/roundtrip-spec.md and the
// reference Python is in .tmp/reference_compare.py.
//
// Modes:
//   - default: assert unexcused diffs are empty per fixture; assert ratchet.
//   - FIDELITY_DISCOVER=1: print the sorted, deduped, '*'-normalized union of
//     computed diff paths and PASS (used to (re)build the ledger).
//   - FIDELITY_KNOWN_GAPS=<path>: read the ledger from <path> instead of the
//     default repo-root manifest (used to self-verify against a scratch copy).
//
// This test never contacts a live server.

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"reflect"
	"sort"
	"strings"
	"testing"
)

const (
	fidelityLang            = "go"
	fixturesDir             = "../test-fixtures/expectations"
	defaultKnownGapsRelPath = "../test-fixtures/expectations/known-gaps.json"
)

// keyToMultiValue fields (headers/params/trailers) accept an object form or an
// [{name,values|value}] array form; both canonicalize to { name -> [values] }.
var multiValueKeys = map[string]bool{
	"headers":               true,
	"queryStringParameters": true,
	"trailers":              true,
}

// keyToValue fields (cookies) canonicalize to { name -> value }.
var singleValueKeys = map[string]bool{
	"cookies": true,
}

// canonMulti canonicalizes a keyToMultiValue field to map name -> []values
// (values are normalized by the caller).
func canonMulti(v interface{}) map[string]interface{} {
	out := map[string]interface{}{}
	switch vv := v.(type) {
	case map[string]interface{}:
		for k, val := range vv {
			if lst, ok := val.([]interface{}); ok {
				out[k] = lst
			} else {
				out[k] = []interface{}{val}
			}
		}
	case []interface{}:
		for _, e := range vv {
			em, ok := e.(map[string]interface{})
			if !ok {
				continue
			}
			name, ok := em["name"]
			if !ok {
				continue
			}
			nameStr, ok := name.(string)
			if !ok {
				continue
			}
			vals, present := em["values"]
			if !present {
				vals = em["value"] // nil if also absent
			}
			if lst, ok := vals.([]interface{}); ok {
				out[nameStr] = lst
			} else {
				out[nameStr] = []interface{}{vals}
			}
		}
	}
	return out
}

// canonSingle canonicalizes a keyToValue field to map name -> value.
func canonSingle(v interface{}) map[string]interface{} {
	out := map[string]interface{}{}
	switch vv := v.(type) {
	case map[string]interface{}:
		for k, val := range vv {
			out[k] = val
		}
	case []interface{}:
		for _, e := range vv {
			em, ok := e.(map[string]interface{})
			if !ok {
				continue
			}
			name, ok := em["name"]
			if !ok {
				continue
			}
			if nameStr, ok := name.(string); ok {
				out[nameStr] = em["value"]
			}
		}
	}
	return out
}

// norm applies NORM (null==absent) and CANON (dual-encoding) in one recursive
// pass keyed by the parent object's key name.
func norm(v interface{}, key string) interface{} {
	if v == nil {
		return nil
	}
	if multiValueKeys[key] {
		out := map[string]interface{}{}
		for k, vs := range canonMulti(v) {
			lst := vs.([]interface{})
			normed := make([]interface{}, len(lst))
			for i, x := range lst {
				normed[i] = norm(x, "")
			}
			out[k] = normed
		}
		return out
	}
	if singleValueKeys[key] {
		out := map[string]interface{}{}
		for k, x := range canonSingle(v) {
			out[k] = norm(x, "")
		}
		return out
	}
	switch vv := v.(type) {
	case map[string]interface{}:
		out := map[string]interface{}{}
		for k, x := range vv {
			if x == nil {
				continue // drop null entries (absent == null)
			}
			out[k] = norm(x, k)
		}
		return out
	case []interface{}:
		out := make([]interface{}, len(vv))
		for i, x := range vv {
			out[i] = norm(x, "")
		}
		return out
	default:
		return vv
	}
}

func rootOr(path string) string {
	if path == "" {
		return "<root>"
	}
	return path
}

func sortedMapKeys(m map[string]interface{}) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

// diffs walks the canonicalized/normalized INPUT (a) against the OUTPUT (b) and
// returns the list of differing path strings. A "[ADDED]" suffix marks a key
// present only in the output.
func diffs(a, b interface{}, path string) []string {
	var res []string
	switch av := a.(type) {
	case map[string]interface{}:
		bv, ok := b.(map[string]interface{})
		if !ok {
			return []string{rootOr(path)}
		}
		for _, k := range sortedMapKeys(av) {
			p := k
			if path != "" {
				p = path + "." + k
			}
			if bval, present := bv[k]; !present {
				res = append(res, p)
			} else {
				res = append(res, diffs(av[k], bval, p)...)
			}
		}
		for _, k := range sortedMapKeys(bv) {
			if _, present := av[k]; !present {
				p := k
				if path != "" {
					p = path + "." + k
				}
				res = append(res, p+" [ADDED]")
			}
		}
		return res
	case []interface{}:
		bv, ok := b.([]interface{})
		if !ok {
			return []string{rootOr(path)}
		}
		for i, v := range av {
			p := fmt.Sprintf("%s.%d", path, i)
			if i >= len(bv) {
				res = append(res, p)
			} else {
				res = append(res, diffs(v, bv[i], p)...)
			}
		}
		return res
	default:
		if !reflect.DeepEqual(a, b) {
			res = append(res, rootOr(path))
		}
		return res
	}
}

func isAllDigits(s string) bool {
	if s == "" {
		return false
	}
	for _, r := range s {
		if r < '0' || r > '9' {
			return false
		}
	}
	return true
}

// star replaces numeric path segments with '*' so a ledger entry is independent
// of a fixture's array lengths.
func star(p string) string {
	parts := strings.Split(p, ".")
	for i, s := range parts {
		if isAllDigits(s) {
			parts[i] = "*"
		}
	}
	return strings.Join(parts, ".")
}

// excused reports whether a ledger entry excuses the given diff path: some entry
// G matches path P iff len(G) <= len(P) and each G segment equals the P segment
// or is '*' matching an all-digit P segment.
func excused(path string, entries []string) bool {
	pseg := strings.Split(path, ".")
	for _, g := range entries {
		gseg := strings.Split(g, ".")
		if len(gseg) > len(pseg) {
			continue
		}
		match := true
		for i := 0; i < len(gseg); i++ {
			if gseg[i] == pseg[i] {
				continue
			}
			if gseg[i] == "*" && isAllDigits(pseg[i]) {
				continue
			}
			match = false
			break
		}
		if match {
			return true
		}
	}
	return false
}

func stripAdded(p string) string {
	return strings.TrimSuffix(p, " [ADDED]")
}

// roundTripDiffs computes the stripped ('[ADDED]' removed) diff paths for one
// fixture: unmarshal into Expectation, re-marshal, then compare as generic JSON.
func roundTripDiffs(t *testing.T, data []byte) []string {
	t.Helper()
	var input interface{}
	if err := json.Unmarshal(data, &input); err != nil {
		t.Fatalf("parse fixture: %v", err)
	}
	// Mirror the client's own round-trip: json.Unmarshal(data, &e) with the
	// error deliberately ignored. encoding/json records the first type
	// mismatch but still populates every other field, so a field whose Go type
	// cannot hold the fixture's JSON shape simply drops out — exactly the
	// fidelity gap this harness measures.
	var e Expectation
	_ = json.Unmarshal(data, &e)
	outBytes, err := json.Marshal(e)
	if err != nil {
		t.Fatalf("marshal Expectation: %v", err)
	}
	var output interface{}
	if err := json.Unmarshal(outBytes, &output); err != nil {
		t.Fatalf("reparse round-tripped Expectation: %v", err)
	}
	raw := diffs(norm(input, ""), norm(output, ""), "")
	paths := make([]string, len(raw))
	for i, p := range raw {
		paths[i] = stripAdded(p)
	}
	return paths
}

func knownGapsPath() string {
	if p := os.Getenv("FIDELITY_KNOWN_GAPS"); p != "" {
		return p
	}
	return defaultKnownGapsRelPath
}

func loadGaps(t *testing.T, lang string) []string {
	t.Helper()
	data, err := os.ReadFile(knownGapsPath())
	if err != nil {
		t.Fatalf("read known-gaps manifest: %v", err)
	}
	var raw map[string]json.RawMessage
	if err := json.Unmarshal(data, &raw); err != nil {
		t.Fatalf("parse known-gaps manifest: %v", err)
	}
	entry, ok := raw[lang]
	if !ok {
		return nil
	}
	var gaps []string
	if err := json.Unmarshal(entry, &gaps); err != nil {
		t.Fatalf("parse %q gap list: %v", lang, err)
	}
	return gaps
}

func loadFixtures(t *testing.T) []string {
	t.Helper()
	matches, err := filepath.Glob(filepath.Join(fixturesDir, "*.json"))
	if err != nil {
		t.Fatalf("glob fixtures: %v", err)
	}
	var files []string
	for _, f := range matches {
		if filepath.Base(f) == "known-gaps.json" {
			continue // the manifest is not a fixture
		}
		files = append(files, f)
	}
	sort.Strings(files)
	if len(files) == 0 {
		t.Fatalf("no fixtures found under %s", fixturesDir)
	}
	return files
}

func TestRoundTripFidelity(t *testing.T) {
	fixtures := loadFixtures(t)

	// Discovery mode: print the sorted, deduped, '*'-normalized union of all
	// computed diff paths and PASS (used to (re)build the ledger).
	if os.Getenv("FIDELITY_DISCOVER") == "1" {
		union := map[string]bool{}
		for _, f := range fixtures {
			data, err := os.ReadFile(f)
			if err != nil {
				t.Fatalf("read fixture %s: %v", f, err)
			}
			for _, p := range roundTripDiffs(t, data) {
				union[star(p)] = true
			}
		}
		out := make([]string, 0, len(union))
		for p := range union {
			out = append(out, p)
		}
		sort.Strings(out)
		rendered, _ := json.MarshalIndent(out, "", "  ")
		fmt.Printf("FIDELITY_DISCOVER go gaps (%d):\n%s\n", len(out), string(rendered))
		return
	}

	gaps := loadGaps(t, fidelityLang)

	// Per-fixture: assert every diff is excused by the ledger.
	// Also record which ledger entries actually excused something (for the ratchet).
	usedEntry := make([]bool, len(gaps))
	for _, f := range fixtures {
		f := f
		t.Run(filepath.Base(f), func(t *testing.T) {
			data, err := os.ReadFile(f)
			if err != nil {
				t.Fatalf("read fixture: %v", err)
			}
			var unexcused []string
			for _, p := range roundTripDiffs(t, data) {
				matched := false
				for i, g := range gaps {
					if excused(p, []string{g}) {
						usedEntry[i] = true
						matched = true
					}
				}
				if !matched {
					unexcused = append(unexcused, p)
				}
			}
			if len(unexcused) > 0 {
				sort.Strings(unexcused)
				t.Errorf("unexcused round-trip fidelity diffs (%d):\n  %s\nAdd these to known-gaps.json[%q] (with '*' for array indices) or fix the model.",
					len(unexcused), strings.Join(unexcused, "\n  "), fidelityLang)
			}
		})
	}

	// Ratchet: every ledger entry must excuse at least one diff across all
	// fixtures. A stale entry means the model was fixed and the entry must be
	// removed.
	t.Run("__ratchet__", func(t *testing.T) {
		var stale []string
		for i, g := range gaps {
			if !usedEntry[i] {
				stale = append(stale, g)
			}
		}
		if len(stale) > 0 {
			sort.Strings(stale)
			t.Errorf("stale known-gaps.json[%q] entries (excuse no diff; remove them):\n  %s",
				fidelityLang, strings.Join(stale, "\n  "))
		}
	})
}
